### PR TITLE
Export Rancher images from system-charts

### PR DESCRIPTION
**Pull request by orangedeng**
**Problem:**
1. We don't export the images from system-charts to rancher-image.txt
but we need them in Rancher release.
2. We don't export the Windows image to rancher-image.txt

**Solution:**
1. Clone system-charts repo and collect the images from it.
2. Export Windows images to `rancher-windows-image.txt` and generate
powershell scripts `rancher-save-images.ps1`, `rancher-load-images.ps1`
and `rancher-mirror-to-rancher-org.ps1`.
3. And we need to remove unuse and not pushed Windows images.

Related issues: 
https://github.com/rancher/rancher/issues/18351
https://github.com/rancher/rancher/issues/18534
System-charts PR: https://github.com/rancher/system-charts/pull/22
Related types PR: https://github.com/rancher/types/pull/746

### CODE REVIEW

**Comment by alena1108** [comment]
@orangedeng just curious - do we have a specific branch on per rancher version?

**Comment by orangedeng** [comment]
@alena1108 we don't have that for now and I am cloning the master branch. But I think need to have version controller for system-charts too.

**Comment by orangedeng** [comment]
After system-charts PR is merged, the CI process will show the external-dns image too.

File: pkg/image/export/main.go [hunk 1]
```diff
@@ -262,3 +278,110 @@ func findStrings(obj map[string]interface{}, found map[string]bool) {
...
+		IconPath:  path,
+		Hash:      "",
+	}
+	index, err := helm.LoadIndex()
+	if err != nil {
+		return nil, err
+	}
+	for k, versions := range index.IndexFile.Entries {
+		var maxVersion *libhelm.ChartVersion
+		for _, version := range versions {
```
Comment (thxCode) [inline pkg/image/export/main.go]: I saw the `libhelm` has already done a reverse sorting for the `versions` array, can we use the `versions[0]` as the maximum version?

File: pkg/image/export/main.go [hunk 2]
```diff
@@ -262,3 +278,110 @@ func findStrings(obj map[string]interface{}, found map[string]bool) {
...
+
+	return rtn, nil
+}
+
+func walkFunc(images map[string]struct{}, versions map[string]string, basePath, path string, info os.FileInfo, err error) error {
+	relPath, err := filepath.Rel(basePath, path)
+	if err != nil {
+		return err
+	}
+	found := false
```
Comment (thxCode) [inline pkg/image/export/main.go]: `var found bool` is clearer

File: pkg/image/export/main.go [hunk 3]
```diff
@@ -262,3 +278,110 @@ func findStrings(obj map[string]interface{}, found map[string]bool) {
...
+		return err
+	}
+	dataInterface := map[interface{}]interface{}{}
+	if err := yaml.Unmarshal(data, &dataInterface); err != nil {
+		return err
+	}
+
+	if err := walkthroughMap(dataInterface, func(inputMap map[interface{}]interface{}) error {
+		return generateImages(inputMap, images)
+	}); err != nil {
```
Comment (thxCode) [inline pkg/image/export/main.go]: I saw there aren't any errors will be returned in `generateImages`, should we make `walkthroughMap` return none error?

File: pkg/image/export/main.go [hunk 4]
```diff
@@ -262,3 +278,110 @@ func findStrings(obj map[string]interface{}, found map[string]bool) {
...
+		return err
+	}
+	return nil
+}
+
+func generateImages(inputMap map[interface{}]interface{}, output map[string]struct{}) error {
```
Comment (thxCode) [inline pkg/image/export/main.go]: There are not any returning errors.

**Comment by orangedeng** [comment]
Updated

**Comment by orangedeng** [comment]
@alena1108 I don't think we can add Windows images into 
`rancher-images.txt` because those bash scripts are using the images in it to pull and push
I would prefer to generate a new images list file `rancher-windows-images.txt` and related Powershell scripts for Windows images.

**Comment by thxCode** [comment]
LGTM

**Comment by alena1108** [comment]
> @alena1108 I don't think we can add Windows images into
> `rancher-images.txt` because those bash scripts are using the images in it to pull and push
> I would prefer to generate a new images list file `rancher-windows-images.txt` and related Powershell scripts for Windows images.

good with me, can you file an issue for that @orangedeng

**Comment by orangedeng** [comment]
@alena1108 Filed a new issue for windows images: https://github.com/rancher/rancher/issues/18534 and rebased the code.

**Pull request merged**